File: SiriusQuality/SiriusQuality-EnergyBalance_Original/src/simplace/SiriusQuality_EnergyBalance/CanopyTemperature.java
package net.simplace.sim.components.SiriusQuality-EnergyBalance;
import  java.io.*;
import  java.util.*;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.LocalDateTime;
import net.simplace.sim.model.FWSimComponent;
import net.simplace.sim.util.FWSimVarMap;
import net.simplace.sim.util.FWSimVariable;
import net.simplace.sim.util.FWSimVariable.CONTENT_TYPE;
import net.simplace.sim.util.FWSimVariable.DATA_TYPE;
import org.jdom2.Element;


public class CanopyTemperature extends FWSimComponent
{
    private FWSimVariable<Double> minTair;
    private FWSimVariable<Double> cropHeatFlux;
    private FWSimVariable<Double> specificHeatCapacityAir;
    private FWSimVariable<Double> conductance;
    private FWSimVariable<Double> lambdaV;
    private FWSimVariable<Double> rhoDensityAir;
    private FWSimVariable<Double> maxTair;
    private FWSimVariable<Double> maxCanopyTemperature;
    private FWSimVariable<Double> minCanopyTemperature;

    public CanopyTemperature(String aName, HashMap<String, FWSimVariable<?>> aFieldMap, HashMap<String, String> aInputMap, Element aSimComponentElement, FWSimVarMap aVarMap, int aOrderNumber)
    {
        super(aName, aFieldMap, aInputMap, aSimComponentElement, aVarMap, aOrderNumber);
    }

    public CanopyTemperature(){
        super();
    }

    @Override
    public HashMap<String, FWSimVariable<?>> createVariables()
    {
        addVariable(FWSimVariable.createSimVariable("minTair", "minimum air temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"degC", 30, 45, 0.7, this));
        addVariable(FWSimVariable.createSimVariable("cropHeatFlux", "Crop heat flux", DATA_TYPE.DOUBLE, CONTENT_TYPE.rate,"g/m**2/d", 0, 10000, 447.912, this));
        addVariable(FWSimVariable.createSimVariable("specificHeatCapacityAir", "Specific heat capacity of dry air", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"MJ/kg/degC", None, None, 0.00101, this));
        addVariable(FWSimVariable.createSimVariable("conductance", "the boundary layer conductance", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"m/d", 0, 10000, 598.685, this));
        addVariable(FWSimVariable.createSimVariable("lambdaV", "latent heat of vaporization of water", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"MJ/kg", 0, 10, 2.454, this));
        addVariable(FWSimVariable.createSimVariable("rhoDensityAir", "Density of air", DATA_TYPE.DOUBLE, CONTENT_TYPE.constant,"kg/m**3", None, None, 1.225, this));
        addVariable(FWSimVariable.createSimVariable("maxTair", "maximum air Temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.input,"degC", 30, 45, 7.2, this));
        addVariable(FWSimVariable.createSimVariable("maxCanopyTemperature", "maximal Canopy Temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"degC", 30, 45, null, this));
        addVariable(FWSimVariable.createSimVariable("minCanopyTemperature", "minimal Canopy Temperature", DATA_TYPE.DOUBLE, CONTENT_TYPE.state,"degC", 30, 45, null, this));

        return iFieldMap;
    }
    @Override
    protected void process()
    {
        double t_minTair = minTair.getValue();
        double t_cropHeatFlux = cropHeatFlux.getValue();
        double t_specificHeatCapacityAir = specificHeatCapacityAir.getValue();
        double t_conductance = conductance.getValue();
        double t_lambdaV = lambdaV.getValue();
        double t_rhoDensityAir = rhoDensityAir.getValue();
        double t_maxTair = maxTair.getValue();
        double t_maxCanopyTemperature = maxCanopyTemperature.getDefault();
        double t_minCanopyTemperature = minCanopyTemperature.getDefault();
        if (t_minTair == (double)(999) && t_maxTair == (double)(-999))
        {
            t_minCanopyTemperature = (double)(999);
            t_maxCanopyTemperature = (double)(-999);
        }
        else
        {
            t_minCanopyTemperature = t_minTair + (t_cropHeatFlux / (t_rhoDensityAir * t_specificHeatCapacityAir * t_conductance / t_lambdaV * 1000.00d));
            t_maxCanopyTemperature = t_maxTair + (t_cropHeatFlux / (t_rhoDensityAir * t_specificHeatCapacityAir * t_conductance / t_lambdaV * 1000.00d));
        }
        maxCanopyTemperature.setValue(t_maxCanopyTemperature, this);
        minCanopyTemperature.setValue(t_minCanopyTemperature, this);
    }

    @Override
    protected void init()
    {
    }
    public HashMap<String, FWSimVariable<?>> fillTestVariables(int aParamIndex, TEST_STATE aDefineOrCheck)
    {
        return iFieldMap;
    }

    @Override
    protected FWSimComponent clone(FWSimVarMap aVarMap)
    {
        return new CanopyTemperature(iName, iFieldMap, iInputMap, iSimComponentElement, aVarMap, iOrderNumber);
    }
}

File: SiriusQuality/SiriusQuality-EnergyBalance_Original/src/cs/SiriusQuality_EnergyBalance/CanopyTemperature.cs
using System;
using System.Collections.Generic;
using System.Linq;
public class CanopyTemperature
{
    private double _specificHeatCapacityAir;
    public double specificHeatCapacityAir
    {
        get { return this._specificHeatCapacityAir; }
        set { this._specificHeatCapacityAir= value; } 
    }
    private double _lambdaV;
    public double lambdaV
    {
        get { return this._lambdaV; }
        set { this._lambdaV= value; } 
    }
    private double _rhoDensityAir;
    public double rhoDensityAir
    {
        get { return this._rhoDensityAir; }
        set { this._rhoDensityAir= value; } 
    }
    /// <summary>
    /// Constructor of the CanopyTemperature component")
    /// </summary>  
    public CanopyTemperature() { }
    
    public void  CalculateModel(EnergyBalanceCompositeState s, EnergyBalanceCompositeState s1, EnergyBalanceCompositeRate r, EnergyBalanceCompositeAuxiliary a, EnergyBalanceCompositeExogenous ex)
    {
        //- Name: CanopyTemperature -Version: 1.0, -Time step: 1
        //- Description:
    //            * Title: CanopyTemperature Model
    //            * Authors: Peter D. Jamieson, Glen S. Francis, Derick R. Wilson, Robert J. Martin
    //            * Reference: https://doi.org/10.1016/0168-1923(94)02214-5
    //            * Institution: New Zealand Institute for Crop and Food Research Ltd.,
    //New Zealand Institute for Crop and Food Research Ltd.,
    //New Zealand Institute for Crop and Food Research Ltd.,
    //New Zealand Institute for Crop and Food Research Ltd.
    //
    //            * ExtendedDescription: It is calculated from the crop heat flux and the boundary layer conductance
    //            * ShortDescription: It is calculated from the crop heat flux and the boundary layer conductance
        //- inputs:
    //            * name: specificHeatCapacityAir
    //                          ** description : Specific heat capacity of dry air
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** max : None
    //                          ** min : None
    //                          ** default : 0.00101
    //                          ** unit : MJ/kg/degC
    //            * name: maxTair
    //                          ** description : maximum air Temperature
    //                          ** inputtype : variable
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DOUBLE
    //                          ** max : 45
    //                          ** min : 30
    //                          ** default : 7.2
    //                          ** unit : degC
    //            * name: cropHeatFlux
    //                          ** description : Crop heat flux
    //                          ** inputtype : variable
    //                          ** variablecategory : rate
    //                          ** datatype : DOUBLE
    //                          ** max : 10000
    //                          ** min : 0
    //                          ** default : 447.912
    //                          ** unit : g/m**2/d
    //            * name: lambdaV
    //                          ** description : latent heat of vaporization of water
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** max : 10
    //                          ** min : 0
    //                          ** default : 2.454
    //                          ** unit : MJ/kg
    //            * name: minTair
    //                          ** description : minimum air temperature
    //                          ** inputtype : variable
    //                          ** variablecategory : auxiliary
    //                          ** datatype : DOUBLE
    //                          ** max : 45
    //                          ** min : 30
    //                          ** default : 0.7
    //                          ** unit : degC
    //            * name: rhoDensityAir
    //                          ** description : Density of air
    //                          ** inputtype : parameter
    //                          ** parametercategory : constant
    //                          ** datatype : DOUBLE
    //                          ** max : None
    //                          ** min : None
    //                          ** default : 1.225
    //                          ** unit : kg/m**3
    //            * name: conductance
    //                          ** description : the boundary layer conductance
    //                          ** inputtype : variable
    //                          ** variablecategory : state
    //                          ** datatype : DOUBLE
    //                          ** max : 10000
    //                          ** min : 0
    //                          ** default : 598.685
    //                          ** unit : m/d
        //- outputs:
    //            * name: maxCanopyTemperature
    //                          ** description : maximal Canopy Temperature
    //                          ** datatype : DOUBLE
    //                          ** variablecategory : state
    //                          ** max : 45
    //                          ** min : 30
    //                          ** unit : degC
    //            * name: minCanopyTemperature
    //                          ** description : minimal Canopy Temperature
    //                          ** datatype : DOUBLE
    //                          ** variablecategory : state
    //                          ** max : 45
    //                          ** min : 30
    //                          ** unit : degC
        double maxTair = a.maxTair;
        double cropHeatFlux = r.cropHeatFlux;
        double minTair = a.minTair;
        double conductance = s.conductance;
        double maxCanopyTemperature;
        double minCanopyTemperature;
        if (minTair == (double)(999) && maxTair == (double)(-999))
        {
            minCanopyTemperature = (double)(999);
            maxCanopyTemperature = (double)(-999);
        }
        else
        {
            minCanopyTemperature = minTair + (cropHeatFlux / (rhoDensityAir * specificHeatCapacityAir * conductance / lambdaV * 1000.00));
            maxCanopyTemperature = maxTair + (cropHeatFlux / (rhoDensityAir * specificHeatCapacityAir * conductance / lambdaV * 1000.00));
        }
        s.maxCanopyTemperature= maxCanopyTemperature;
        s.minCanopyTemperature= minCanopyTemperature;
    }
}